include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,483 @@
+    static_assert(sizeof...(_consts) == _n, "the amount of _consts must be equal to _n");
+    static_assert(_r > 0, "_r must be more than 0");
+    static_assert(_w > 0 && _w <= ::std::numeric_limits<scalar_type>::digits,
+                  "_w must satisfy 0 < _w < ::std::numeric_limits<_UIntType>::digits");
+    static_assert(::std::numeric_limits<scalar_type>::digits <= 64,
+                  "size of the scalar _UIntType (in case of sycl::vec<T, N> the size of T) must be less than 64 bits");
+    static_assert(::std::is_unsigned_v<scalar_type>, "_UIntType must be unsigned type or vector of unsigned types");
+
+    static constexpr ::std::array<scalar_type, array_size> multipliers =
+        internal::experimental::get_even_array_from_tuple<scalar_type>(::std::make_tuple(_consts...),
+                                                                       ::std::make_index_sequence<array_size>{});
+    static constexpr ::std::array<scalar_type, array_size> round_consts =
+        internal::experimental::get_odd_array_from_tuple<scalar_type>(::std::make_tuple(_consts...),
+                                                                      ::std::make_index_sequence<array_size>{});
+    static constexpr scalar_type
+    min()
+    {
+        return 0;
+    }
+    static constexpr scalar_type
+    max()
+    {
+        return ::std::numeric_limits<scalar_type>::max() & in_mask;
+    }
+    static constexpr scalar_type default_seed = 20111115u;
+
+    /* Constructors and seeding functions */
+    philox_engine() : philox_engine(default_seed) {}
+    explicit philox_engine(scalar_type __seed) { seed(__seed); }
+    void
+    seed(scalar_type __seed = default_seed)
+    {
+        seed_internal({__seed & in_mask});
+    }
+
+    /* Set the state to arbitrary position */
+    void
+    set_counter(const ::std::array<scalar_type, word_count>& __counter)
+    {
+        auto __end = __counter.end();
+        for (::std::size_t __i = 0; __i < word_count; __i++)
+        {
+            // all counters are set in reverse order
+            state_.X[__i] = (*--__end) & in_mask;
+        }
+    }
+
+    /* Generating functions */
+    result_type
+    operator()()
+    {
+        result_type __ret = generate_internal<internal::type_traits_t<result_type>::num_elems>();
+        return __ret;
+    }
+    /* operator () overload for result portion generation */
+    result_type
+    operator()(unsigned int __random_nums)
+    {
+        result_type __ret = generate_internal<internal::type_traits_t<result_type>::num_elems>(__random_nums);
+        return __ret;
+    }
+
+    /* Shift the counter only forward relative to its current position */
+    void
+    discard(unsigned long long __z)
+    {
+        discard_internal(__z);
+    }
+
+    /* Equality operators */
+    friend bool
+    operator==(const philox_engine& __x, const philox_engine& __y)
+    {
+        if (!::std::equal(__x.state_.X.begin(), __x.state_.X.end(), __y.state_.X.begin()) ||
```

Comment:
this can be simplified to:
return 
std::equal(__x.state_.X.begin(), __x.state_.X.end(), __y.state_.X.begin()) &&
std::equal(__x.state_.K.begin(), __x.state_.K.end(), __y.state_.K.begin()) &&
std::equal(__x.state_.Y.begin(), __x.state_.Y.end(), __y.state_.Y.begin()) &&
__x.state_.idx != __y.state_.idx;

Reply:
Thanks for the proposal, done! Only a minor remark - the last statement should be `_x.state.idx == _y.state.idx;`